[CORE-13363] container:priority_queue

BenPope opened the pull request:
Add:
* `async_while` - an asynchronous while loop for large ranges.
* `priority_queue` - like `std::priority_queue` with move semantics.
* `bounded_priority_queue` - keeps the top-k elements.
* `async_merge_top_k` - efficiently merge a range of sorted range.

Both of the queues support:
* `async_push_range` - to avoid reactor stalls
* `async_extract_sorted() &&` - obtain a sorted container without reactor stalls.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

BenPope committed:
chunked_vector: Satisfy random_access_iterator concepts

The iterators are already random_access, but don't meet the
requirements for the concepts.

* Add an alias for `iterator_concept`
* Remove cv_ref from `value_type`
* Fix `pointer` and `reference` to add back cv_ref

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
container/priority_queue: A replacement for std::priority_queue.

priority_queue is mostly a drop-in for std::prioprity_queue:
* Supports move semantics
* Supports reserve
* pop() moves and returns top value

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
container: bounded_priority_queue

Keeps the top-k of values pushed to it.

Signed-off-by: Ben Pope <ben@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--12937a11-38f1-4727-930b-3ff70b8fb42c-->
<!--TEST RESULTS-->
<!--ca3d747a-2dcc-4609-a584-5490b851ffe2-->
<!--TEST RESULTS-->
<!--11d5799f-4c6e-4886-9c16-241daec5cafb-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74156">build#74156</a></summary>


| test_class                          | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:------------------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| QuickTerminateTest                  | test_terminate       | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74156#0199e2ec-2910-4547-86e5-0fb0898ccbd0 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                  |
| RandomNodeOperationsTest            | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74156#0199e2ec-2911-4141-a305-2ba6cdca3fa5 | FLAKY         | 20/21    | upstream reliability is '98.22695035460993'. current run reliability is '95.23809523809523'. drift is 2.98886 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations      |
| WriteCachingFailureInjectionE2ETest | test_crash_all       | {"use_transactions": false}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/74156#0199e2ec-2915-4cc7-a66f-5cc1478ecb13 | FLAKY         | 15/21    | upstream reliability is '82.79835390946502'. current run reliability is '71.42857142857143'. drift is 11.36978 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74268">build#74268</a></summary>


| test_class                          | test_method                                  | test_arguments                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------------------|:---------------------------------------------|:-----------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FollowerFetchingTest                | test_with_leadership_transfers               | {"fetch_from": "fetch-from-cloud-topic"} | integration | https://buildkite.com/redpanda/redpanda/builds/74268#0199e8ef-3cae-4532-842d-f313332401af | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FollowerFetchingTest&test_method=test_with_leadership_transfers        |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74268#0199e8ef-3cad-49c3-a36f-34dc54d2e2cc | FLAKY         | 19/21    | upstream reliability is '90.68062827225131'. current run reliability is '90.47619047619048'. drift is 0.20444 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74268#0199e8f2-706f-4035-bd04-247d97fe57d0 | FLAKY         | 19/21    | upstream reliability is '90.68062827225131'. current run reliability is '90.47619047619048'. drift is 0.20444 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}              | integration | https://buildkite.com/redpanda/redpanda/builds/74268#0199e8f2-706f-4035-bd04-247d97fe57d0 | FLAKY         | 15/21    | upstream reliability is '82.02511773940346'. current run reliability is '71.42857142857143'. drift is 10.59655 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74386">build#74386</a></summary>


| test_class                      | test_method              | test_arguments                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowIndexingGlobalConfig      | test_overrides_remove    | null                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74386#0199f1ab-16f5-477a-8052-9e1b26bda75a | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowIndexingGlobalConfig&test_method=test_overrides_remove         |
| DatalakeTransactionTests        | test_with_transactions   | {"cloud_storage_type": 1, "compaction": false}                    | integration | https://buildkite.com/redpanda/redpanda/builds/74386#0199f1b9-183f-4d47-91c6-81b92692d24e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeTransactionTests&test_method=test_with_transactions          |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false}                | integration | https://buildkite.com/redpanda/redpanda/builds/74386#0199f1ab-16f5-477a-8052-9e1b26bda75a | FLAKY         | 16/21    | upstream reliability is '94.97206703910615'. current run reliability is '76.19047619047619'. drift is 18.78159 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| TimeQueryTest                   | test_timequery           | {"batch_cache": false, "cloud_storage": true, "spillover": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74386#0199f1ab-16f9-4c53-9f55-530ba4eb542d | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery                             |
</details>
<!--/body-->

pgellert commented on src/v/container/priority_queue.h:
```diff
@@ -220,5 +230,163 @@ class priority_queue : public detail::priority_queue_base<T, Container, Comp> {
+        value_type result = std::move(base::_cont.back());
+        base::_cont.pop_back();
+        return result;
+    }
```
As we discussed, I think we won't need these `top()` and `pop()` and so I think we should remove them as they feel overly expensive and counter to the way `bounded_priority_queue` is supposed to be used to implement an async top K algorithm.

I think what we need instead is an `ss::future<> extract_sorted()` (or `extract_top_k()` or similar naming) method that calls `base::pop_heap()` `base::_cont.size()` times to sort the underlying container and move it out. If I am following the code correctly, calling `base::pop_heap()` will just work because it will repeatedly move the `max` (as per `bounded_priority_queue::Comp`) value to the end of the vector and so it will sort the elements in `bounded_priority_queue::Comp` order.

This should be async to avoid stalls if K/capacity is large.

BenPope replied:
> As we discussed, I think we won't need these `top()` and `pop()` and so I think we should remove them as they feel overly expensive and counter to the way `bounded_priority_queue` is supposed to be used to implement an async top K algorithm.

Yes, the cost is really bad.

> I think what we need instead is an `ss::future<> extract_sorted()` (or `extract_top_k()` or similar naming) method that calls `base::pop_heap()` `base::_cont.size()` times to sort the underlying container and move it out. If I am following the code correctly, calling `base::pop_heap()` will just work because it will repeatedly move the `max` (as per `bounded_priority_queue::Comp`) value to the end of the vector and so it will sort the elements in `bounded_priority_queue::Comp` order.

Yes, this works, and I have it on my branch.

> This should be async to avoid stalls if K/capacity is large.

Agreed, I have an `async_extract_sorted` on my branch.

pgellert commented on src/v/container/tests/priority_queue_test.cc:
```diff
@@ -0,0 +1,189 @@
+        results.push_back(pq.pop());
+    }
+
+    EXPECT_TRUE(std::ranges::is_sorted(results, std::ranges::less{}));
```
nit: tests should have no logic if possible, so I would replace this with a check on the explicit values
```
EXPECT_THAT(results, ElementsAre(
        this->make_value(1),
        this->make_value(1),
        this->make_value(3),
        this->make_value(4),
        this->make_value(5)
    ));
```

pgellert commented on src/v/container/tests/priority_queue_test.cc:
```diff
@@ -0,0 +1,189 @@
+TYPED_TEST_SUITE(StdCompatibilityTest, StdCompatibilityTypes);
+
+// Test drop-in replacement compatibility with std::priority_queue
+TYPED_TEST(StdCompatibilityTest, StdCompatibility) {
+    std::priority_queue<typename TestFixture::value_type> std_pq;
+    priority_queue<typename TestFixture::value_type> our_pq;
+
+    for (int i : {3, 1, 4, 1, 5}) {
+        std_pq.push(this->make_value(i));
+        our_pq.push(this->make_value(i));
+    }
+
+    // Test size and empty
+    EXPECT_EQ(std_pq.size(), our_pq.size());
+    EXPECT_EQ(std_pq.empty(), our_pq.empty());
+
+    // Test top and pop equivalence
+    while (!std_pq.empty() && !our_pq.empty()) {
+        EXPECT_EQ(std_pq.top(), our_pq.top());
+        std_pq.pop();
+        std::ignore = our_pq.pop();
+    }
+
+    EXPECT_EQ(std_pq.empty(), our_pq.empty());
+}
```
nice test! I'm wondering if it's worth making this a random test that takes a random (0--100) elements of random int values as inputs and does this check and running that through `bazel test //... --runs_per_test=100` in case it highlights something.

pgellert commented on src/v/container/priority_queue.h:
```diff
@@ -0,0 +1,224 @@
+            if (input_size == 1) {
+                base::push_heap();
+            } else if (input_size > 1) {
+                base::make_heap();
```
Is `make_heap()` safe to call when the containers (either the input range or the existing `base::_cont`) are large wrt. reactor stalls?

If not, I'd be tempted to drop `push_range` and force the caller to use `ssx::for_each_async` with `push(...)`.

BenPope replied:
> Is `make_heap()` safe to call when the containers (either the input range or the existing `base::_cont`) are large wrt. reactor stalls?

> If not, I'd be tempted to drop `push_range` and force the caller to use `ssx::for_each_async` with `push(...)`.

How would you feel about a about an additional async version of the function, `async_push_range`?

pgellert replied:
Yeah that sounds good to me. We're going to integrate the insertions here: https://github.com/redpanda-data/redpanda/blob/8cda5945f3c13bc77a4d3007a1b0f181543c5196/src/v/redpanda/admin/services/broker.cc#L105-L123

So we could either use the existing `ssx::for_each_async` and call `push(...)` right after the protobuf object is created, or we could build a temporary `chunked_vector` and `async_push_range`.

pgellert reviewed (commented):
Looks great! I think the `bounded_priority_queue` API needs some minor adjustments to better optimize for the async top-K use case and to make sure that this is free from reactor stalls, but I think the building blocks are all there so it's looking good.

BenPope committed:
container/priority_queue: Implement async operations

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
ssx: Introduce async_merge_top_k

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
container/priority_queue: Implement benchmark

```
test               iterations      median         mad         min         max      allocs       tasks        inst      cycles
chunked_pq.small         7864   108.849us    97.496ns   102.978us   109.851us    1017.000       1.035   1317336.8    571404.7
chunked_pq.reduce          42     4.845ms    50.980us     4.790ms     5.128ms    1017.000      13.429  55144184.2  26038707.7
chunked_pq.large            2   645.818ms     5.639ms   636.184ms   661.591ms 1001011.000    1305.000 2044356083.2 3548639685.1
```

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
ssx/async_algorithm: Implement async_while

Signed-off-by: Ben Pope <ben@redpanda.com>

StephanDollberg commented on src/v/ssx/async_algorithm.h:
```diff
@@ -347,4 +347,27 @@ async_for_each_counter(async_counter& counter, Container&& container, Fn f) {
+ss::future<> async_while(Counter counter, Cond cond, Fn f) {
+    while (cond()) {
+        f();
+        ++counter.count;
```
Here I think you want an inner while loop that runs `f()` up to `remaining<Traits>` times. Similar to how `async_for_each_coro` has the inner nested `for_each`.

BenPope replied:
> Here I think you want an inner while loop that runs `f()` up to `remaining<Traits>` times. Similar to how `async_for_each_coro` has the inner nested `for_each`.

The condition needs checking on every iteration.

StephanDollberg replied:
Yeah but you want the inner loop to be tight so you want something like:

```
while (cond()) {
        int i = 0;
        for (; i < Traits::interval && cond(); ++i) {
            f();
        }
        counter.count += i;
        if (counter.count >= Traits::interval) {
            ... 
```

BenPope replied:
Done

StephanDollberg commented on src/v/container/tests/priority_queue_bench.cc:
```diff
@@ -0,0 +1,148 @@
+    source_data.reserve(InputSize);
+
+    for (size_t j = 0; j < InputSize; ++j) {
+        source_data.emplace_back();
```
Is the emplace back empty here intentionally?

BenPope replied:
> Is the emplace back empty here intentionally?

Yes, the default constructor of `value<1>` is sufficient. I can't use any kind of fill, since `value<1>` is move-only.

StephanDollberg replied:
Right, I guess I don't fully understand the motivation behind the `value` struct given we just use it with `<1>` anyway. Why can't it just be an integer?

Anyway, why I am bringing this up is that for microbench stability it would be best if you create the `source_data` data once in a fixture and then use it from there as I don't see a need to move all the elements into the priority queue. Like the microbench should focus on testing the priority-queue behaviour and not whether copying or moving its `T`s is expensive.

BenPope replied:
Rewritten.

StephanDollberg replied:
Sorry for causing trouble here, I do see now that the container bench is actually partly bugged.

In the benchframe work the fixture is not getting recreated for each repetition (as one might expect) so for test functions that modify the fixture members (sort, pop etc) one still needs to explicitly copy the members first.

Btw if this is needed for some other work/blocking some DR work etc then feel free to merge as is and fix it up in a separate PR.

BenPope replied:
> Sorry for causing trouble here, I do see now that the container bench is actually partly bugged.
> 
> In the benchframe work the fixture is not getting recreated for each repetition (as one might expect) so for test functions that modify the fixture members (sort, pop etc) one still needs to explicitly copy the members first.

After pushing, I did notice that it''s running too fast to be doing the right thing!

> Btw if this is needed for some other work/blocking some DR work etc then feel free to merge as is and fix it up in a separate PR.

The fix will look like a pretty bad regression!

pgellert commented on src/v/container/priority_queue.h:
```diff
@@ -0,0 +1,408 @@
+
+    auto max_element(this auto&& self) {
+        return std::ranges::max_element(self._cont, self.comp());
+    }
```
nit: `max_element` is unused now
```suggestion
```

pgellert commented on src/v/container/priority_queue.h:
```diff
@@ -0,0 +1,408 @@
+
+namespace detail {
+
+template<typename Comp>
+struct invert_comparator {
+    template<typename T, typename U>
+    constexpr bool operator()(T&& a, U&& b) const {
+        return _underlying(std::forward<U>(b), std::forward<T>(a));
+    }
+
+    [[no_unique_address]] Comp _underlying;
+};
```
I haven't come across `[[no_unique_address]]` before. What does it achieve here?

BenPope replied:
> I haven't come across `[[no_unique_address]]` before. What does it achieve here?

Most of the time `Comp` is stateless, since `invert_comparator` has no other state, it doesn't need a memory address, or a register, and that allows it to be inlined more easily, and generally reduce overhead.

pgellert commented on src/v/container/tests/priority_queue_bench.cc:
```diff
@@ -0,0 +1,148 @@
+            co_await pq.async_push_range(std::forward<T>(input));
+            auto heap = co_await std::move(pq).async_extract_sorted();
+            co_return heap;
+        };
```
It looks like clang tidy is complaining about these corouting lambdas here

BenPope replied:
> It looks like clang tidy is complaining about these corouting lambdas here

Yeah, I'm [doing it the right way](https://github.com/scylladb/seastar/blob/master/doc/lambda-coroutine-fiasco.md#solution-c23-and-up), too. Easy workaround:
```c++
        // NOLINTNEXTLINE(cppcoreguidelines-avoid-capturing-lambda-coroutines)
        return [output_max]<typename T>(
```

pgellert reviewed (commented):
Looks great! 🔥 Clang tidy is failing, but otherwise this looks good to go

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/0c969852e07b3548ee0f9951911c79aed983a159..8b62efa2fafd63c1b886b87c48448d852ec9483b)
* Improve inner-loop of `async_while`
* Improved test coverage
* Fixed the implementation of `extract_async_sorted`.
* Rewrote the benchmark

BenPope commented on src/v/ssx/async_algorithm.h:
```diff
@@ -347,4 +347,63 @@ async_for_each_counter(async_counter& counter, Container&& container, Fn f) {
+  typename Traits = async_algo_traits,
+  std::predicate Cond,
+  std::invocable Fn>
+ss::future<> async_while_counter(async_counter& counter, Cond cond, Fn f) {
```
Done
